Pre-Lecture HW

1.

In [1]:
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)
df.isna().sum()

row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64

2.

In [7]:
import pandas as pd

# Load the CSV file from the provided URL
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)

# Get the number of rows and columns
num_rows, num_columns = df.shape

print(f"The dataset has {num_rows} rows and {num_columns} columns.")


The dataset has 391 rows and 11 columns.


Observation: An observation refers to a single data point or data entry in a dataset. Each observation has its own row, which includes different kinds of information about it. In terms of the Animal Crossing dataset, each observation is a different villager, and each row contains information on its personality and characteristics.

Variable: A variable is a single attribute that provides information on each observation. Variables are set as columns. In the Animal Crossing dataset, the different variables describe each villager(observation). The characteristics include the villagers' name, gender, birthday, and more.

3.

In [8]:
import pandas as pd

# Load the dataset from the URL
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)

# 1. General Information about the Dataset
print("Dataset Information:")
df.info()
print("\n")

# 2. Summary Statistics for Numerical Columns
print("Summary Statistics for Numerical Columns:")
print(df.describe())
print("\n")

# 3. Summary Statistics for Categorical Columns
print("Summary Statistics for Categorical Columns:")
print(df.describe(include='object'))
print("\n")

# 4. Count of Unique Values in Each Column
print("Count of Unique Values in Each Column:")
print(df.nunique())
print("\n")

# 5. Checking for Missing Values
print("Count of Missing Values in Each Column:")
print(df.isnull().sum())


Dataset Information:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 391 entries, 0 to 390
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   row_n        391 non-null    int64 
 1   id           390 non-null    object
 2   name         391 non-null    object
 3   gender       391 non-null    object
 4   species      391 non-null    object
 5   birthday     391 non-null    object
 6   personality  391 non-null    object
 7   song         380 non-null    object
 8   phrase       391 non-null    object
 9   full_id      391 non-null    object
 10  url          391 non-null    object
dtypes: int64(1), object(10)
memory usage: 33.7+ KB


Summary Statistics for Numerical Columns:
            row_n
count  391.000000
mean   239.902813
std    140.702672
min      2.000000
25%    117.500000
50%    240.000000
75%    363.500000
max    483.000000


Summary Statistics for Categorical Columns:
             id     name gender speci

4.

In [4]:
import pandas as pd

# Load the dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

# Get the number of rows and columns
print("Shape of the dataset:", df.shape)

# Get the default description (numerical columns)
print("\nNumerical Summary:")
print(df.describe())


Shape of the dataset: (891, 15)

Numerical Summary:
         survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200


a) The df.shape function returns the total number of rows and columns in a given dataset. In this Titanic dataset, it returned (891, 15), which means there are 891 rows and 15 columns. However, the df.describe() function will return statistics only on numerical columns, which are columns that include data types int and float. As we can see, out of the 15 total columns in the dataset, df.describe() only returned statistics on 6 of them. The other 9 columns not shown are not numerical, but of a different type such as categorical.

b) The 'count' value represents how many non-missing entries there are for each column. For example, the 'survived' column has a count value of 891, which is equal to the number of rows in the dataset. However, the count value for the 'age' column is only 714, meaning there are 177 missing entries for this column.The count value helps to spot which columns contain missing data.

5.

Attribute: Attributes are similar to variables, in that they store values and information of specific objects. These values include data such as strings, numbers, and lists. Attributes do not end in () and do not take arguments because they can be accessed directly.

Method: Methods are functions which perform actions on specific objects. Unlike attributes, methods need to be called upon to perform operations, so they end in () and take arguments. 

Post-Lecture HW

6.

Count: The number of non-missing data entries per column in a dataset.

Mean: The average of all values in a column. The mean is calculated by adding all the values in a column together, then dividing the sum by the total number of values.

Std (Standard Deviation): The average spread or variance in a dataset. 

Min: The smallest value of a column in a dataset.

25%: The value under which 25% of data entries in a column are found.

50%: The value in which 50% of data entries are found over and 50% of data entries are found under.

75%: The value under which 75% of data entries in a column are found.

Max: The largest value of a column in a dataset.

7.1

A case in which using df.dropna() might be peferred over using del df['col'] is when there is a dataset with missing values, and you'd like to only work with complete data. To do this, you would have to remove all rows with missing values so that you can be left with rows without any missing values. Using df.dropna() will do exactly that and remove any row with a missing value. Del df['col'] will only delete a specific column regardless of whether or not it has missing values, and there may still be missing values remaining in other rows. Using it will not help achieve the goal of working only with complete data.

7.2

A case in which using del df['col'] might be peferred over using df.dropna() is when there is a dataset with multiple columns, and one specific column contains too many missing values for it to be of any use. del df['col'] is able to remove the column so that you are left with all other columns and can continue to use the dataset effectively. Using df.dropna() will remove all rows with missing values, which is not desired in this case. Many columns containing useful data will shrink due to losing the majority of their rows, and it defeats the purpose of trying to keep all other columns intact.

7.3

It can be useful to use del df['col'] before using df.dropna() to minimize the number of data entries lost. When del df['col'] is used first, you can hand pick any columns with large amounts of missing values to delete first. Using df.dropna() after will clean up any remaining missing values and leave you with as many valid data entries as possible. This order helps to avoid deleting entire rows of data, when only a small number of entries in that row were missing.

7.4

In [11]:
import pandas as pd

# Load the dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

# Display the shape of the dataset
print("Shape of the dataset:", df.shape)

# Display all the columns in the dataset
print("\nColumns in the dataset:", df.columns.tolist())

# Display a description of the dataset
print("\nDescription of the dataset:")
print(df.describe(include='all'))


Shape of the dataset: (891, 15)

Columns in the dataset: ['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare', 'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town', 'alive', 'alone']

Description of the dataset:
          survived      pclass   sex         age       sibsp       parch  \
count   891.000000  891.000000   891  714.000000  891.000000  891.000000   
unique         NaN         NaN     2         NaN         NaN         NaN   
top            NaN         NaN  male         NaN         NaN         NaN   
freq           NaN         NaN   577         NaN         NaN         NaN   
mean      0.383838    2.308642   NaN   29.699118    0.523008    0.381594   
std       0.486592    0.836071   NaN   14.526497    1.102743    0.806057   
min       0.000000    1.000000   NaN    0.420000    0.000000    0.000000   
25%       0.000000    2.000000   NaN   20.125000    0.000000    0.000000   
50%       0.000000    3.000000   NaN   28.000000    0.000000    0.000000   
75%     

Above is the dataset before I remove all missing values. There are 15 columns and 891 rows.

In [12]:
del df['deck']

In [13]:
del df['age']

In [26]:
df.isna().sum()

survived       0
pclass         0
sex            0
sibsp          0
parch          0
fare           0
embarked       2
class          0
who            0
adult_male     0
embark_town    2
alive          0
alone          0
dtype: int64

In [29]:
# Remove all rows with any missing values
df_cleaned = df.dropna()

# Display the shape of the cleaned DataFrame
print("Shape of the cleaned dataset (after removing rows with missing values):", df_cleaned.shape)

# Verify that there are no missing values left
print("\nMissing values per column after cleanup:\n", df_cleaned.isnull().sum())


Shape of the cleaned dataset (after removing rows with missing values): (889, 13)

Missing values per column after cleanup:
 survived       0
pclass         0
sex            0
sibsp          0
parch          0
fare           0
embarked       0
class          0
who            0
adult_male     0
embark_town    0
alive          0
alone          0
dtype: int64


As seen above, there are now no missing values in this dataset. My approach to removing all missing values was to first remove the 'deck' and 'age' columns, as they had the most missing values out of all the other columns. After removing them, I decided to remove the rows of the last two missing values. 

I removed the data in this order to minimize the loss of valid data points. If I had used df.dropna() before removing any columns, it would have deleted every row with a missing value, which would leave me with at most 203 rows and only 23% of my data left. By removing the 'deck' and 'age' columns first, I keep the majority of my data. After removing these columns, I was left with just two rows with missing values. I had the option to remove the 'embarked' and 'embark_town' columns, however that would mean I delete 889 valid data entries each. Instead, I can save much more of my data by deleting only the two rows with missing values.

After removing all missing values, the dataset now has 13 columns and 889 rows.

8.1

df.groupby("col1")["col2"].describe() sorts the dataset into groups by unique values in one column, and then tells descriptive details (such as mean, std, max) for a second column within each group.

In this Titanic dataset, I can use this function to group the data by sex, and get further details on the fares paid by each sex.

In [32]:
df.groupby("sex")["fare"].describe()

count       mean        std   min        25%   50%    75%       max
sex                                                                        
female  314.0  44.479818  57.997698  6.75  12.071875  23.0  55.00  512.3292
male    577.0  25.523893  43.138263  0.00   7.895800  10.5  26.55  512.3292

By analyzing this data, we can see that the average female had to pay a substantially higher fare than the average male. The max fare for both sexes were the same, however some males were able to board for free while all females had to pay. The 50th and 75th percentile for females were also more than double that for males.

Further analysis of this data may be able to provide deep insight on social constructs, sexism, the economy, and more from the time of The Titanic.

8.2

df.describe() will give counts for all non-missing values in the entire dataset. df.groupby("col1")["col2"].describe() will only include the counts for non-missing values in 'col1', and will be divided into the groups that col1 was split into. It will exclude counts from col2.

8.3.A

In [1]:
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)
df.isna().sum()

NameError: name 'pd' is not defined

For this mistake, I definitely found it easier to fix this issue with ChatGPT. When I presented this error, ChatGPT immediately told me what was wrong, how I can fix the issue, a visual example of the issue, and common issues to check when encountering this error. Everything was well organized and presented in a way that is easy to understand. If I encountered this issue, I think the ChatGPT explanation would help me resolve it fairly quickly and leave me without any questions.

Google search is a much different story. The first few results are forum and blog  posts with people publicly asking how to fix this error. The user responses to these posts give quite vague instructions and it is difficult to find a proper solution to this. Compared to asking ChatGPT, it is harder to fix this error by searching on Google.

8.3.B

In [2]:
import pandas as pd
url = "titanics.csv"
df = pd.read_csv(url)
df.isna().sum()

FileNotFoundError: [Errno 2] No such file or directory: 'titanics.csv'

Nearly the exact same results from ChatGPT and Google as 8.3.A. For this error, ChatGPT told me to verify the file path, make sure the file exists, correct any typos, and check the working dictionary. As for Google search, the results were forum and blog posts with unclear solutions. ChatGPT provides better troubleshooting help again.

8.3.C

In [3]:
import pandas as pd
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)
df.isna().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [2]:
DF.groupby("sex")["age"].describe()

NameError: name 'DF' is not defined

For this mistake, ChatGPT and Google were both equally helpful. Results from both mentioned how python is case sensitive, and that it is important to avoid typos and define a varaible correctly. They both included visual examples of the error and solution. Either one of them could effectively help somebody fix this error.

8.3.D

In [6]:
df.isna.sum

AttributeError: 'function' object has no attribute 'sum'

In this case, ChatGPT is back to being the much more helpful tool. It explained how this error could come up in different scenariors, such as variable name shadowing and an incorrect use of parantheses. It again showed visual examples of the error and solution, and said that it is important to not misuse parantheses when calling a function. As for Google, the majority of results were forums with unclear advice.

8.3.E

In [7]:
df.group_by("sex")["age"].describe()

AttributeError: 'DataFrame' object has no attribute 'group_by'

ChatGPT immediately tells me that the correct spelling for the function is groupby and not group_by, and shows how to use the function correctly. Google yet again only results in forums with many clueless people asking why this happens, with no clear cut solution given by any other users.

8.3.F

In [9]:
df.groupby("SEX")["AGE"].describe()

KeyError: 'SEX'

For this error, neither ChatGPT or Google were very helpful. Google really had nothing to provide, as most people who asked on forums did not receive any answers, or the answers users gave would not exactly fix my exact error. ChatGPT also did not know the exact error or solution, but it did mention using valid parameters. However, that was not its main solution as it mostly talked about how to group columns.

8.3.G

In [10]:
df.groupby(sex)["age"].describe()

NameError: name 'sex' is not defined

Yet again, ChatGPT provides far more helpful answers than Google. ChatGPT quickly tells me to the mistake is either because of missing quotations or an undefined variable. Google search results in more forums where users are saying the method to fix this is by using more complicated functions and adding them together. Their answers are confusing and not what I am looking for.

9.

Yes.

Summary for the chatbot session used for questions 1-6:

Summary of This Chatbot Session
In this session, we covered the following topics:

Dataset Exploration:

You downloaded the "Animal Crossing characters" dataset from a URL and wanted to know the number of rows and columns it contains. We discussed how to use Python and the pandas library to find this information.
We explored the meanings of "observations" (rows) and "variables" (columns) in the context of this dataset.
You also asked how to create simple summaries of each column in a dataset using Python, for which I provided a script utilizing the pandas functions such as info(), describe(), and others.
Differences Between DataFrame Functions:

We discussed why there may be a difference in the number of columns reported by df.shape and df.describe(). df.shape shows the total number of rows and columns, while df.describe() only provides summaries for numerical columns by default.
I explained the difference in count values found in the summary generated by df.describe(), highlighting that it represents the number of non-missing values for each column.
Attributes vs. Methods in Python:

You wanted to understand the difference between an attribute function (does not end in ()) and a method function (ends in ()).
Attributes represent data or properties of an object, while methods are functions associated with an object that perform actions or operations.
Explanation of Statistical Terms:

I provided explanations for common statistics (count, mean, std, min, 25%, 50%, 75%, max) generated by the pandas library, detailing what each statistic represents and its purpose in data analysis.
This session provided an overview of data exploration and basic statistics using Python, which can help you analyze datasets more effectively.


https://chatgpt.com/c/66e0a4fd-bbe4-800c-a38d-a5184bc72b25





Summary of the chatbot session used for questions 7 and 8:

Summary of Our Exchanges
Difference Between df.dropna() and del df['col']:

We discussed the use cases for when df.dropna() is preferred (removing rows with missing values across the entire DataFrame) versus del df['col'] (deleting a specific column from the DataFrame).
Combining del df['col'] and df.dropna():

We explained how using del df['col'] before df.dropna() can be helpful to remove irrelevant or problematic columns that contain missing data, making the subsequent row-wise removal (dropna) more targeted and effective.
Loading the Titanic Dataset:

I assisted you with loading the Titanic dataset from a URL using pandas, and we went over how to check the shape of the dataset, display all columns, and provide a description of the dataset.
Removing Missing Data:

We went over how to remove rows with missing data using df.dropna() and how to display the description of the cleaned dataset.
Using groupby with describe():

We explained the purpose of df.groupby("col1")["col2"].describe() and provided an example to illustrate how it groups data by one column and provides descriptive statistics for another.
Differences in count Values with Missing Data:

We covered why count values differ when using df.describe() versus df.groupby("col1")["col2"].describe() in the presence of missing values.
Fixing Common Python Errors:

NameError: name 'pd' is not defined: Explained how to fix this error by properly importing pandas (import pandas as pd).
FileNotFoundError: We discussed how to handle missing file errors by ensuring the correct file path and checking for the existence of the file.
NameError: name 'DF' is not defined: We explained how this is likely due to a typo or incorrect capitalization and showed how to correctly define and reference variables.
AttributeError: 'function' object has no attribute 'sum': We covered how this error can happen due to name shadowing or incorrect usage of functions and provided examples to resolve it.
AttributeError: 'DataFrame' object has no attribute 'group_by': We pointed out the correct method is .groupby(), not .group_by().
TypeError("You have to supply one of 'by' and 'level'"): We explained how to fix this error by correctly providing either the by or level parameter when using groupby.
NameError: name 'sex' is not defined: We described how this error could occur due to missing quotes around column names or undefined variables and provided solutions to fix it.
This summary should cover all the key points we discussed and provide a clear record of our interactions to include in your homework assignment. If you need any more details or clarifications, feel free to ask!


https://chatgpt.com/c/66e201f0-926c-800c-acd4-82cda406c664